### PR TITLE
Add ref-spec matching of spaceros repository (#91).

### PR DESCRIPTION
**Pull request by xfiderek**
## Description
This PR fixes https://github.com/space-ros/docker/issues/91, by matching  [spaceros](https://github.com/space-ros/space-ros/) repo branch to current local branch during build. Assuming that current local branch is `feature/foo`, the behavior is as follows:

- If `SPACEROS_GIT_REF` is **not** supplied by user via CLI (default behavior):
    - If `feature/foo` exists in spaceros repo, ros2.repos from that branch will be used
    - Otherwise file will be cloned from `main` branch
- If user overrides `SPACEROS_GIT_REF` via CLI
    - If that branch exists in spaceros repo, that branch will be used
    - otherwise earthly returns error

It is also now possible to specify different url for spaceros git repository, which can be useful in testing scenarios, when contributors are forking spaceros repo. 
If `SPACEROS_REPO_URL` is not supplied by user, then we use wget to clone ros2.repos file from spaceros repo. Otherwise we use Earthly's `GIT CLONE` command to get the file (see discussion below).

Check updated readme file for more details.

## Summary of changes
* Added logic described above to Earthfile
* Updated README.md

## How changes were tested
I have developed python file to test the changes: https://gist.github.com/xfiderek/a697dd33739c035ac6b2b1d6cdee8af1

### CODE REVIEW

File: spaceros/Earthfile [hunk 1]
```diff
@@ -94,9 +102,18 @@ setup:
 
 sources:
   FROM +setup
-  COPY github.com/space-ros/space-ros:earthly-wrapper+repos-file/ros2.repos ros2.repos
+  RUN mkdir spaceros_repo_dir
+
+  IF[ -z "$(git ls-remote ${SPACEROS_REPO_URL} ${GIT_BRANCH})" ]
+    RUN echo "branch '${GIT_BRANCH}' does not exist in spaceros repo, cloning main branch"
+    GIT CLONE ${SPACEROS_REPO_URL} spaceros_repo_dir/
```
[the remark this answers is not in the document]
Comment (xfiderek) [reply spaceros/Earthfile:109]: These are good points. Let me break it down:

- *"Can we at least make a third case that detects if we're just using the main branch of the default repo and uses the githubusercontent CDN to grab the file?"*  - 
- For default spaceros repo, I would even go one step further and say that we should use CDN whenever `SPACEROS_REPO_URL="https://github.com/space-ros/space-ros.git"`, regardless of branch. I think that it is important, because can imagine that if we have multiple spaceros distros, we would rarely pull from main (you would rather pull from specific tag)
- *"I guess a more complete solution would be checking the url for common host platforms like GitHub and GitLab which expose files directly under a specific URL but that might be out of scope for this issue."* -> yes, but here we would ideally need to handle cases when such repos are private. Earthly's `GIT CLONE` handles that automagically (check [earthly docs](https://docs.earthly.dev/docs/guides/best-practices#git-clone-vs-run-git-clone))

In summary, I would refactor that so that we use CDN whenever we pull from default spaceros repo, and use GIT CLONE otherwise. I will document that in the code so that its clear for future generations :D.

Sounds good to you @Bckempa?

**Comment by xfiderek** [comment]
> Not a huge deal, but while I see it executing the feedback line:
> 
> ```
> +sources | *cached* --> RUN echo "cloning branch '${GIT_BRANCH}' from spaceros repo"
> ```
> 
> I don't see the actual echo with the substituted output. Can we find a better way to make sure it's clear in logs what branch was used?
> 
> Either way, seems to work though earthly's caching is so near magical that I want to play with it more before we merge, but you got my green checkmark.

I think i will just remove these `echo` statements as they bring unnecessary noise. I could not find any way to log something to stdout using earthly. Is it ok with you?

File: spaceros/Earthfile [hunk 2]
```diff
@@ -14,14 +14,23 @@
...
 FROM ubuntu:jammy
 
+# current git branch, prefilled by earthly
+# https://docs.earthly.dev/docs/earthfile/builtin-args
+ARG EARTHLY_GIT_BRANCH
+
+# these can be overriden from cli. Example:
+# `earthly +image --SPACEROS_REPO_URL=https://...`
+ARG SPACEROS_REPO_URL="https://github.com/space-ros/space-ros.git"
+ARG GIT_BRANCH=${EARTHLY_GIT_BRANCH}
```
[the remark this answers is not in the document]
Comment (xfiderek) [reply spaceros/Earthfile:24]: 1. Yes, that makes sense. I will change the name to `SPACEROS_GIT_REF`
2. Thanks for bringing it up as it reminded me that we have an ambiguity in https://github.com/space-ros/space-ros repo. We have both branch `humble-2023.10.0` and tag with the same name. In such cases, tools like git clone would prefer branches (and I could not find easy way to change that behavior). I suggest we should delete that branch from the repo and avoid such things in the future.

File: spaceros/Earthfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (xfiderek) [reply spaceros/Earthfile:109]: No worries, I figured it out, I will post the PR today in the evening

**Comment by xfiderek** [comment]
Ok @Bckempa, now it should work as expected. Please take a look again. 
To make sure that corner cases are covered, I ended up writing python file to make sure that all works as expected. I post it above. I also added you to the repository, and as soon as you accept you will be able to run this file locally.

File: spaceros/Earthfile [hunk 3]
```diff
@@ -92,9 +91,40 @@ setup:
...
+
+  # main purpose of this command is to make sure that the git ls-remote results are not cached
+  RUN --no-cache echo "Cloning spaceros repo artifacts"
+
+  # current git branch, prefilled by earthly: https://docs.earthly.dev/docs/earthfile/builtin-args
+  ARG EARTHLY_GIT_BRANCH
+
+  ARG SPACEROS_REPO_URL="https://github.com/space-ros/space-ros.git"
+  # if current local branch does not exist in target repo then use main. note that branch supplied from CLI overrides this behavior.
+  ARG SPACEROS_GIT_REF="$( [ -n \"$(git ls-remote $SPACEROS_REPO_URL $EARTHLY_GIT_BRANCH)\" ] && echo $EARTHLY_GIT_BRANCH || echo 'main' )"
```
Comment (xfiderek) [inline spaceros/Earthfile:107]: As we want to re-run ls-remote every time we can no longer keep them at the top of the file.

**Comment by EzraBrooks** [comment]
looks like this should be good to go after a rebase

**Comment by xfiderek** [comment]
Rebase applied
